### PR TITLE
Notice by package

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Add a new notice reporter that reports licenses by package. See the commit messages for details.

### CODE REVIEW

**Comment by mnonnenmacher** [comment]
I still need to improve the test case in the last commit, but the rest can already be reviewed.

File: reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                val licensesWithCopyrights = licenses.associateWith { license ->
+                    // Remove copyrights which are already contained in the license file.
+                    val allCopyrights = (licenseFindingsMap[license].orEmpty() + copyrightsInFile).toSet()
+                    val processedCopyrights = CopyrightStatementsProcessor().process(allCopyrights)
+                    val remainingProcessedCopyrights =
+                        processedCopyrights.processedStatements.filterNot { (_, sources) ->
+                            copyrightsInFile.any { it in sources }
+                        }
+                    val remainingUnprocessedCopyrights =
+                        processedCopyrights.unprocessedStatements - copyrightsInFile
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt]: In conjunction with `remainingUnprocessedCopyrights` the name `processedCopyrights` might be a bit confusing.
Consider `processCopyrightsResult` instead so that you would end up with `processCopyrightsResult.processed...` and `processCopyrightsResult.unprocessed...

File: reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        noticeReport: NoticeReport,
+        licenseTextProvider: LicenseTextProvider,
+        copyrightGarbage: CopyrightGarbage
+    ) {
+        val archiver = config.scanner?.archive?.createFileArchiver() ?: FileArchiver(
+            LICENSE_FILENAMES,
+            LocalFileStorage(DEFAULT_ARCHIVE_DIR)
+        )
+
+        noticeReport.findings.forEach { (id, licenseFindingsMap) ->
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt]: This code block starting at `line 70` is rather large. 
It does multiple things at once and I think it'd become easier to maintain
if (at least) the following two things were separated
1. create a model class for that report
2. Turn that data class into a string representation

e.g. as data class I'd imagine something like the following plus additional things needed in the report:
```
data class PackageEntry(val licensesTexts: List<String>,  val packageName: String)
```

**Comment by mnonnenmacher** [comment]
> I still need to improve the test case in the last commit, but the rest can already be reviewed.

The test case is still not finished, I need to replace the test project with a smaller one.

File: utils/src/main/kotlin/storage/FileArchiver.kt [hunk 3]
```diff
@@ -65,6 +67,19 @@ class FileArchiver(
...
+            storage.read(getArchivePath(storagePath)).use { input ->
+                input.unpackZip(directory)
+            }
+            true
+        } catch (e: IOException) {
+            false
+        }
+    }
+
+    private fun getArchivePath(storagePath: String) = "${storagePath.removeSuffix("/")}/$ARCHIVE_FILE_NAME"
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply utils/src/main/kotlin/storage/FileArchiver.kt]: I do not find the existing code hard to read.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 4]
```diff
@@ -23,6 +23,30 @@ import java.io.File
 import java.security.MessageDigest
 import java.util.EnumSet
 
+private fun List<String>.addUppercaseAndCapitalized() = flatMap { listOf(it, it.toUpperCase(), it.capitalize()) }
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt]: Done.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 5]
```diff
@@ -23,6 +23,30 @@ import java.io.File
...
+    "patents"
+).addUppercaseAndCapitalized()
+
+/**
+ * A list of globs that match files that often define the root license of a project, but are no license files and are
+ * therefore not contained in [LICENSE_FILENAMES].
+ */
+val ROOT_LICENSE_FILENAMES = listOf(
+    "readme*"
+).addUppercaseAndCapitalized()
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt]: Currently it is required because readme files shall never be included in notice files. But the "patents" argument is correct, maybe adapting the docs would help? I'll think about it on Monday.

File: reporter/src/main/kotlin/reporters/AbstractNoticeReporter.kt [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+import com.here.ort.model.config.CopyrightGarbage
+import com.here.ort.model.licenses.LicenseConfiguration
+import com.here.ort.reporter.LicenseTextProvider
+import com.here.ort.reporter.Reporter
+import com.here.ort.reporter.ResolutionProvider
+import com.here.ort.utils.ScriptRunner
+
+import java.io.OutputStream
+
+abstract class AbstractNoticeReporter : Reporter {
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/AbstractNoticeReporter.kt]: Not sure, at this time we do not make any promises regarding API stability anyway.

File: utils/src/main/kotlin/storage/FileArchiver.kt [hunk 7]
```diff
@@ -65,6 +67,19 @@ class FileArchiver(
...
     }
+
+    fun unarchive(directory: File, storagePath: String): Boolean {
+        return try {
+            storage.read(getArchivePath(storagePath)).use { input ->
+                input.unpackZip(directory)
+            }
+            true
+        } catch (e: IOException) {
+            false
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply utils/src/main/kotlin/storage/FileArchiver.kt]: I have added the typical exception logging.

File: utils/src/main/kotlin/storage/FileArchiver.kt [hunk 8]
```diff
@@ -65,6 +67,19 @@ class FileArchiver(
             }
         }
 
-        storage.write("${storagePath.removeSuffix("/")}/$ARCHIVE_FILE_NAME", zipFile.inputStream())
+        storage.write(getArchivePath(storagePath), zipFile.inputStream())
     }
+
+    fun unarchive(directory: File, storagePath: String): Boolean {
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply utils/src/main/kotlin/storage/FileArchiver.kt]: Somehow this function felt too long for me to be an expression function, especially after adding the logging above. But I can change it if you want.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 9]
```diff
@@ -40,6 +40,7 @@ import com.here.ort.model.ScannerDetails
 import com.here.ort.model.ScannerRun
 import com.here.ort.model.config.ScannerConfiguration
 import com.here.ort.model.mapper
+import com.here.ort.spdx.LICENSE_FILENAMES
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply scanner/src/main/kotlin/LocalScanner.kt]: Somehow this felt like a good place for these constants, because it already has the `NON_LICENSE_FILENAMES` list which is somehow related. But I'm find with moving this back to utils (but not inside the `FileMatcher` then), as at least currently nothing in the spdx-utils module depends on it.

File: reporter/src/main/kotlin/LicenseTextProvider.kt [hunk 10]
```diff
@@ -24,8 +24,7 @@ package com.here.ort.reporter
  */
 interface LicenseTextProvider {
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/LicenseTextProvider.kt]: Done.

File: reporter/src/main/kotlin/DefaultLicenseTextProvider.kt [hunk 11]
```diff
@@ -32,4 +33,12 @@ class DefaultLicenseTextProvider(private val customLicenseTextsDir: File? = null
             customLicenseTextsDir = customLicenseTextsDir
         )
     }
+
+    override fun hasLicenseText(licenseId: String): Boolean {
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/DefaultLicenseTextProvider.kt]: Done. Also added a commit to convert the function above for consistency.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 12]
```diff
@@ -145,13 +146,20 @@ inline fun <reified T : Enum<T>> enumSetOf(vararg elems: T): EnumSet<T> =
...
     if (id.startsWith("LicenseRef-")) {
-        getLicenseTextFromResource(id) ?: customLicenseTextsDir?.let { getLicenseTextFromDirectory(id, it) }
+        getLicenseTextResource(id)?.readText() ?: customLicenseTextsDir?.let { getLicenseTextFile(id, it)?.readText() }
     } else {
         SpdxLicense.forId(id)?.text ?: SpdxLicenseException.forId(id)?.text?.takeIf { handleExceptions }
     }
 
-private fun getLicenseTextFromResource(id: String): String? =
-    object {}.javaClass.getResource("/licenserefs/$id")?.readText()
+fun hasLicenseText(id: String, handleExceptions: Boolean = false, customLicenseTextsDir: File? = null): Boolean =
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt]: I have removed the redundancy by adding an abstraction for the different license text sources. However, that means more code of course. To me it makes sense to have this function to not have to load the license texts into memory multiple times when it is not really required.

File: reporter/src/funTest/kotlin/reporters/NoticeByPackageReporterTest.kt [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import io.kotlintest.specs.WordSpec
+
+import java.io.ByteArrayOutputStream
+import java.io.File
+
+private fun generateReport(
+    ortResult: OrtResult,
+    copyrightGarbage: CopyrightGarbage = CopyrightGarbage(),
+    postProcessingScript: String? = null
+) =
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/funTest/kotlin/reporters/NoticeByPackageReporterTest.kt]: Done.

File: reporter/src/main/kotlin/Reporter.kt [hunk 14]
```diff
@@ -59,9 +60,11 @@ interface Reporter {
      * The report may be post-processed by a [postProcessingScript] before it is written to [outputStream] whereas the
      * [licenseConfiguration] is passed as a parameter to that script.
      */
+    @Suppress("LongParameterList")
     fun generateReport(
         outputStream: OutputStream,
         ortResult: OrtResult,
+        config: OrtConfiguration = OrtConfiguration(),
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/Reporter.kt]: I plan to replace that anyway by making the `OrtConfiguration` globally available in a separate PR.

File: reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+        archiveDir: File,
+        licenseFindingsMap: LicenseFindingsMap
+    ): MutableMap<String, LicenseFindingsMap> {
+        val matcher = FileMatcher(LICENSE_FILENAMES)
+        val licenseFiles = mutableMapOf<String, LicenseFindingsMap>()
+
+        archiveDir.walkTopDown().forEach { file ->
+            val relativePath = archiveDir.toPath().relativize(file.toPath())
+            if (matcher.matches(relativePath.toString())) {
+                val licenses = findLicensesForFile(scanResult, relativePath)
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt]: I have tested this with large projects and did not find any performance issues. Maybe you can propose a good test project offline on Monday.

File: utils/src/main/kotlin/storage/FileArchiver.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply utils/src/main/kotlin/storage/FileArchiver.kt]: I'm using the term `Consider` usually as an indicator that I'd like the altrnative to be considered and at same time that I'm also fine with it as it is and that I don't request any change with that comment.

We haven't agreed on such indicators as a team, I think we should. What do you think ?

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply scanner/src/main/kotlin/LocalScanner.kt]: I'll leave it up to you. I thought I'd share this because when I moved it out of SPDX one of the reasons was that I could not see how it relates to SPDX.

File: utils/src/main/kotlin/storage/FileArchiver.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/storage/FileArchiver.kt]: How about using a *prefix* like "Alternatively: " or "Optional: " in such cases, similar to like we already use "Nit: "?

[the remark this answers is not in the document]
Comment (fviernau) [reply utils/src/main/kotlin/storage/FileArchiver.kt]: @sschuberth I like both of your proposals, in particular also the `colon`.

File: reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt (hunk 2 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt]: I have split this function into several smaller ones. But I agree it makes sense to further optimize the code, maybe as you proposed. However, I would prefer to do it in a separate PR.

File: reporter/src/main/kotlin/Reporter.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply reporter/src/main/kotlin/Reporter.kt]: Ideally, the addition of this parameter would have been done in a preparing commit, so that this commit really only add the new notice reporter implementation, like the commit message says.

File: reporter/src/funTest/assets/NPM-is-windows-1.0.2-expected-NOTICE [hunk 16]
```diff
@@ -1,4 +1,5 @@
-This project contains or depends on third-party software components pursuant to the following licenses:
+This project contains or depends on third-party software components.
```
Comment (sschuberth) [inline reporter/src/funTest/assets/NPM-is-windows-1.0.2-expected-NOTICE]: Ideally, the code changes that cause the expected test results for the legacy notice reporter to change would have been done in a separate commit, so that this commit really only add the implementation of the new notice reporter. Maybe these changes could have even been squash into the change that separates out the abstract notice reporter.

Comment (sschuberth) [reply reporter/src/funTest/assets/NPM-is-windows-1.0.2-expected-NOTICE]: But *why* does this wording *have* to change after all?

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 17]
```diff
@@ -23,6 +23,30 @@ import java.io.File
...
+    "patents"
+).generateCapitalizationVariants()
+
+/**
+ * A list of globs that match files that often define the root license of a project, but are no license files and are
+ * therefore not contained in [LICENSE_FILENAMES].
+ */
+val ROOT_LICENSE_FILENAMES = listOf(
+    "readme*"
+).generateCapitalizationVariants()
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxUtils.kt]: What was the outcome of you thinking about merging these lists?

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 18]
```diff
@@ -157,14 +160,46 @@ inline fun <reified T : Enum<T>> enumSetOf(vararg elems: T): EnumSet<T> =
...
+    override fun read() = file.readText()
+}
+
+private class SpdxLicenseTextReader(val id: SpdxLicense) : LicenseTextReader() {
+    override fun read() = id.text
+}
+
+private class SpdxExceptionTextReader(val id: SpdxLicenseException) : LicenseTextReader() {
+    override fun read() = id.text
+}
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxUtils.kt]: Wow, these sealed classes are quite some code explosion... how about simply returning a function instead? Like:

```
private fun getLicenseTextReader(
    id: String,
    handleExceptions: Boolean = false,
    customLicenseTextsDir: File? = null
): (() -> String)? =
    if (id.startsWith("LicenseRef-")) {
        getLicenseTextResource(id)?.let { url -> { url.readText() } }
            ?: customLicenseTextsDir?.let { getLicenseTextFile(id, it)?.let { file -> { file.readText() } } }
    } else {
        SpdxLicense.forId(id)?.let { license -> { license.text } }
            ?: SpdxLicenseException.forId(id)?.takeIf { handleExceptions }?.let { exception -> { exception.text } }
    }
```

File: spdx-utils/src/main/kotlin/SpdxUtils.kt [hunk 19]
```diff
@@ -157,14 +160,46 @@ inline fun <reified T : Enum<T>> enumSetOf(vararg elems: T): EnumSet<T> =
  * license text is retrieved from that directory if and only if the license text is not known by ORT.
  */
 fun getLicenseText(id: String, handleExceptions: Boolean = false, customLicenseTextsDir: File? = null): String? =
+    getLicenseTextReader(id, handleExceptions, customLicenseTextsDir)?.read()
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/SpdxUtils.kt]: When following my suggestion to return a function below, use `invoke()` instead of `read()` here.

File: reporter/src/main/kotlin/DefaultLicenseTextProvider.kt [hunk 20]
```diff
@@ -25,11 +25,10 @@ import com.here.ort.spdx.getLicenseText
 import java.io.File
 
 class DefaultLicenseTextProvider(private val customLicenseTextsDir: File? = null) : LicenseTextProvider {
-    override fun getLicenseText(licenseId: String): String? {
-        return getLicenseText(
+    override fun getLicenseText(licenseId: String): String? =
```
Comment (sschuberth) [inline reporter/src/main/kotlin/DefaultLicenseTextProvider.kt]: Commit message: "expression"

File: reporter/src/main/kotlin/reporters/AbstractNoticeReporter.kt [hunk 21]
```diff
@@ -83,6 +84,7 @@ abstract class AbstractNoticeReporter : Reporter {
     override fun generateReport(
         outputStream: OutputStream,
         ortResult: OrtResult,
+        config: OrtConfiguration,
```
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/AbstractNoticeReporter.kt]: Done.

File: reporter/src/main/kotlin/Reporter.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/Reporter.kt]: Done.

File: reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt (hunk 1 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt]: Done.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt (hunk 18 above)
Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt]: Done.

File: reporter/src/main/kotlin/DefaultLicenseTextProvider.kt (hunk 20 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/DefaultLicenseTextProvider.kt]: Done.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt (hunk 19 above)
Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt]: Done.

**Comment by mnonnenmacher** [comment]
@sschuberth @fviernau : Most comments are addressed, I will address the remaining ones tomorrow.

File: spdx-utils/src/main/kotlin/SpdxUtils.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxUtils.kt]: Ping, just as a reminder.

**Comment by mnonnenmacher** [comment]
Replaced by #2044.
